### PR TITLE
New algorithm closest points with new strategies

### PR DESCRIPTION
**Pull request by vissarion**
This PR implements the closest points algorithm of PR https://github.com/boostorg/geometry/pull/707 with new umbrella strategies. 

It contains only the cases of cartesian point - geometry (no box) to keep it small. The rest of the functionality will follow in new PRs. At the end of this process I will close https://github.com/boostorg/geometry/pull/707

### CODE REVIEW

**Comment by vissarion** [comment]
Thanks for the detailed review. I addressed all the comments. The main change is that closest_points will not support legacy strategies. I think this is OK for new algorithms @barendgehrels what do you think?

File: include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+            >::apply(strategies.distance(point, range));
+        
+        auto it_pair = point_to_point_range::apply(point, 
+                                                   boost::begin(range),
+                                                   boost::end(range), 
+                                                   comparable_distance,
+                                                   cd_min);
+        
+        auto closest_point = strategies.closest_points(point, range)
+            .apply(point, *it_pair.first, *it_pair.second);
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp:130]: Minor, we usually use `it_pair->first` and `it_pair->second`

**Review by barendgehrels — CHANGES_REQUESTED** [review]
Cool that we have this now.
I think some of the docs are not in this PR, could you add them?
I added a few comments, can you address them if you agree?
What is the relation from `closest_points` to `closest_feature` - do we need to keep both?

File: include/boost/geometry/algorithms/detail/closest_points/implementation.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include <boost/geometry/algorithms/detail/distance/implementation.hpp>
+#include <boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/multipoint_to_geometry.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/linear_to_linear.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/linear_or_areal_to_areal.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/linear_to_box.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/geometry_to_segment_or_box.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/segment_to_segment.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/segment_to_box.hpp>
+//#include <boost/geometry/algorithms/detail/closest_points/box_to_box.hpp>
```
[the remark this answers is not in the document]
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/closest_points/implementation.hpp:23]: I will uncomment/implement those in the next PR(s). It is just convenient for me to leave it there, but if you insist I will remove it.

[the remark this answers is not in the document]
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/closest_points/implementation.hpp:23]: Processing it in next PR is fine for me.

**Comment by barendgehrels** [comment]
> Thanks for the detailed review. I addressed all the comments. The main change is that closest_points will not support legacy strategies. I think this is OK for new algorithms @barendgehrels what do you think?

I think it's fine, for new algorithms, to not support legacy strategies. We're moving on.

File: include/boost/geometry/algorithms/detail/closest_points/interface.hpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+\tparam Geometry1 \tparam_geometry
+\tparam Geometry2 \tparam_geometry
+\tparam Segment Any type fulfilling a Segment Concept
+\param geometry1 \param_geometry
+\param geometry2 \param_geometry
+\param shortest_seg Output segment containing the closest points
+\qbk{
+[heading Example]
+[closest_points]
+[closest_points_output]
```
[the remark this answers is not in the document]
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/closest_points/interface.hpp:202]: If this is not a problem, I would like to add the example at the end when the full functionality is implemented.

File: test/algorithms/closest_points/common.hpp [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                    Strategy const& strategy,
+                    bool const& swap_geometries,
+                    bool const& default_strategy)
+{
+    compute_result(geometry1, geometry2, exp_resulting_segment, strategy,
+                   default_strategy);
+    if (swap_geometries)
+    {
+        //TODO: enable
+        // swap input geometries and expected segment
```
[the remark this answers is not in the document]
Comment (vissarion) [reply test/algorithms/closest_points/common.hpp:148]: I am not sure I understand the question here, this is for testing the cases with swapped arguments.

**Comment by vissarion** [comment]
> Cool that we have this now. 
>I think some of the docs are not in this PR, could you add them? 

Ι prefer to complete the documentation in the last PR of `closest_points` sequel. 

>I added a few comments, can you address them if you agree? 

Thanks! I answered to all of them.

> What is the relation from `closest_points` to `closest_feature` - do we need to keep both?

`closest_feature` is more like a set of utilities that compute the distance and the closest feature between ranges and geometries. They are used by `distance` and `closest_points` algorithms.

**Comment by barendgehrels** [comment]
> Ι prefer to complete the documentation in the last PR of `closest_points` sequel.

Sure, that's fine, I'll approve, and thanks for processing and answering.

**Review by barendgehrels — APPROVED** [review]

File: test/algorithms/closest_points/common.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (barendgehrels) [reply test/algorithms/closest_points/common.hpp:148]: I referred to the `TODO`. Probably it's already done.

[the remark this answers is not in the document]
Comment (vissarion) [reply test/algorithms/closest_points/common.hpp:147]: Yes, I removed the comment, thanks.

File: include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp (hunk 1 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp:130]: Note that here `it_pair` is a pair of iterators not an iterator of a pair.

Comment (awulkiew) [reply include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp:130]: If the name can be confusing maybe we could change it? This code is taken from `distance()` but here the type is not explicitly stated the only identifier that describes this variable in the code is its name. For me this either looks like an iterator to pair or a range represented as pair of iterators but is neither of these things. It's a representation of a segment. We already have names like `referring_segment` and `pointing_segment` in the code. My point is that the name should probably contain the word `segment`.

Comment (vissarion) [reply include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp:130]: ok I agree, what about `closest_segment` ? It is the closest segment in `range` to the query `point`.

Comment (awulkiew) [reply include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp:130]: `closest_segment` looks good.

**Pull request merged**